### PR TITLE
Task fine tuner

### PR DESCRIPTION
**Pull request by oz03-hub**
Features:
- data extraction scripts
- fine tuning suites
- sbatch scripts to use trainers on unity
- report on decisions made an challenges: https://docs.google.com/document/d/17t-MRulBXyp-WsPVg_WaYRnvEutDUfsqRNVWgjVVFww/edit

Goal:
- Providing a starter point for future research on geomodel fine tuning and annotation integration
- Note limitations discovered during development
- Document the new code

How to test:
I suggest testing fine tuning and data extraction separately, at the end of the steps you have the option to do fine tuning with your own annotations.

**Fine tuning**
- on unity ondemand, start an interactive vscode session for 3 hours with gpu
- in the unity session clone the repository to your Personal space, not the pi space, work in your personal space or you have to change the paths
- switch to Task-fine-tuner branch, follow along set up instructions listed in the readme
- once you have pretrained models, and original training data, you can stop following along the readme,
- Activate your sinr_icml environment so you can start fine-tuning `python fine_tune_main.py`, an example.csv is provided to act as a sample annotation data.
- You can change parameters, such as lr, in fine_tune_main to observe different effects on training
- During training it will print loss at x frequency, if you see the loss decreasing it means it is learning - though the loss values can be unstable depending on how you configure weight parameters, but out of the box if you see changes in loss it means it is learning
- Use `python viz_map.py --name fine_tuned --model_path /path/to/inaturalist-sinr/fine-tuned/demo/${YOUR_SAVED_MODEL_NAME}.pt --taxa_id ${TAXA}` to display predictions for a given taxa, fill the values to the corresponding values
- You can try 5165, 12345, 5174, and 5177 as examples of taxa_id
- For species that were not annotated, their predictions will be the same, you can check it by comparing an original model to your fine tuned model, `python viz_map.py --name original --model_path pretrained_models/model_an_full_input_enc_sin_cos_hard_cap_num_per_class_1000.pt`  will give predictions for 130714, you can run the earlier command with taxa_id=130714 to get fine tuned predictions, they will be the same.
- Details about fine tuning steps can be found [here](https://github.com/UMassCDS/inaturalist-sinr/tree/TASK-fine-tuner?tab=readme-ov-file#fine-tuning)

**Data extraction**

- Firstly, play around with iNatAtor on its main branch and save a few annotations to the database
- To test `data_extraction.ipynb`, clone the repo to the computer that is running your database instance and enter database secrets to .env
- Create the sinr_icml environment in conda for the new computer
- Select the sinr_icml environment for the `data_extraction.ipynb` and start running the notebook, it will run step by step explaining what is going on
- If the csv file with a datetime has been generated it means it was successful.

If you want to use the generated csv for fine tuning, you have to upload the file to unity using unity dashboard
Details for data extraction can be found [here](https://github.com/UMassCDS/inaturalist-sinr/tree/TASK-fine-tuner?tab=readme-ov-file#fine-tuning)

Issues linked:
- https://github.com/orgs/UMassCDS/projects/13/views/1?pane=issue&itemId=70513709
- https://github.com/orgs/UMassCDS/projects/13/views/1?pane=issue&itemId=72130985

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -191,6 +191,23 @@ Now run `python reproduce.py` to create recon results from pretrained models. Wh
 
 Contact Ozel with errors if any.
 
+### 🎛️ Fine Tuning
+Make sure you have followed the steps for downloading training data and pretrained models described above.
+
+To extract annotation data from iNatAtor, use `data_extraction.ipynb` or `data_extraction.py` (to submit as jobs). You will need to confidure a .env that has connection secrets. If your database is running in your local computer and you want to use unity to extract data as a job you will need to build a tunnel. For development and testing I suggest running this script on the same computer as your database and then upload the .csv file into unity to `inaturalist-sinr/data/annotation`.
```
Comment (angelazhu2) [inline README.md]:
```suggestion
To extract annotation data from iNatAtor, use `data_extraction.ipynb` or `data_extraction.py` (to submit as jobs). You will need to configure a .env that has connection secrets. If your database is running in your local computer and you want to use unity to extract data as a job you will need to build a tunnel. For development, I suggest running this script on the same computer as your database and then upload the .csv file into unity to `inaturalist-sinr/data/annotation`.
```

File: README.md [hunk 2]
```diff
@@ -191,6 +191,23 @@ Now run `python reproduce.py` to create recon results from pretrained models. Wh
 
 Contact Ozel with errors if any.
 
+### 🎛️ Fine Tuning
```
Comment (angelazhu2) [inline README.md:199]: actually can we turn this into bullet points? 

1. Extract Data: To extract data do this []. otherwise there's a sample.csv at `inaturalist-sinr/data/annotation`
2. Configure Parameters: 
3. Setup a unity environment to run jobs
4. run the fine tuning job

**Review by angelazhu2 — APPROVED** [review]
great job!! Tested it on unity, seemed to work

**Review by cheerstopaula — APPROVED** [review]

File: fine_tune.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        - taxa: list of taxa values in data
+        - hex_types: denotes absence or presence in a corresponding location
+    """
+
+    print('\nLoading ' + annotation_file)
+    data = pd.read_csv(annotation_file, index_col=0)
+
+    # remove outliers
+    num_obs = data.shape[0]
+    data = data[((data['latitude'] <= 90) & (data['latitude'] >= -90) & (data['longitude'] <= 180) & (data['longitude'] >= -180) )]
```
Comment (ginic) [inline fine_tune.py:38]: Whoa! It's so funny that this has to be done as a data cleaning step. I wonder what's going on with the input data that this is necessary. (No changes needed, just expressing my shock and horror!)

**Review by ginic — APPROVED** [review]
Nice work, @oz03-hub ! The code looks good to me and I feel really confident in everything after the in-person code review. That meeting was really helpful in understanding all the changes and experiments you did. Good idea setting that up!

**Comment by oz03-hub** [comment]
@all merging after 3 approvals
